lib/src/rules/use_build_context_synchronously.dart

```diff
@@ -0,0 +1,239 @@
+  static final _testDirectories = [
+    '${path.separator}test${path.separator}',
+    '${path.separator}integration_test${path.separator}',
+    '${path.separator}test_driver${path.separator}',
+    '${path.separator}testing${path.separator}',
+  ];
+
+  /// Flag to short-circuit `inTestDir` checking when running tests.
+  final bool inTestMode;
+
+  UseBuildContextSynchronously({this.inTestMode = false})
+      : super(
+            name: 'use_build_context_synchronously',
+            description: _desc,
+            details: _details,
+            group: Group.errors,
+            maturity: Maturity.experimental);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var unit = context.currentUnit.unit;
+    if (inTestMode || !inTestDir(unit)) {
+      final visitor = _Visitor(this);
+      registry.addMethodInvocation(this, visitor);
+      registry.addInstanceCreationExpression(this, visitor);
+      registry.addFunctionExpressionInvocation(this, visitor);
+    }
+  }
+
+  static bool inTestDir(CompilationUnit unit) {
+    var path = unit.declaredElement?.source.fullName;
+    return path != null && _testDirectories.any(path.contains);
+  }
+}
+
+class _AwaitVisitor extends RecursiveAstVisitor {
+  bool hasAwait = false;
+
+  @override
+  void visitAwaitExpression(AwaitExpression node) {
+    hasAwait = true;
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  static const _nameBuildContext = 'BuildContext';
+  final Uri _uriFramework;
+
+  final LintRule rule;
+
+  _Visitor(this.rule)
+      : _uriFramework = Uri.parse('package:flutter/src/widgets/framework.dart');
+
+  bool accessesContext(ArgumentList argumentList) {
+    for (var argument in argumentList.arguments) {
+      var argType = argument.staticType;
+      var isGetter = argument is Identifier &&
+          argument.staticElement is PropertyAccessorElement;
+      if (isBuildContext(argType, skipNullable: isGetter)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void check(AstNode node) {
+    // Walk back and look for an async gap that is not guarded by a mounted
+    // property check.
+    AstNode? child = node;
+    while (child != null && child is! FunctionBody) {
+      var parent = child.parent;
+      if (parent is Block) {
+        var statements = parent.statements;
+        var index = statements.indexOf(child as Statement);
+        for (var i = index - 1; i >= 0; i--) {
+          var s = statements.elementAt(i);
+          if (isMountedCheck(s)) {
+            return;
+          } else if (isAsync(s)) {
+            rule.reportLint(node);
+            return;
+          }
+        }
+      } else if (parent is IfStatement) {
+        // if (mounted) { ... do ... }
+        if (isMountedCheck(parent, positiveCheck: true)) {
+          return;
+        }
+      }
+
+      child = parent;
+    }
+  }
+
+  bool isAsync(Statement statement) {
```

Comment:
Do we need to worry about the case where a method or function is invoked that returns a `Future` but where the returned `Future` isn't awaited?

Reply:
It's a good question.  In general I think this will get caught by `unawaited_futures` but something to think about.